intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,525 @@
+
+# Coerces the last two dims to 'H' and 'W'. In Python 2, use the string '...' instead of ...
+print(imgs.refine_names(..., 'H', 'W').names)
+
+def catch_error(fn):
+    try:
+        fn()
+    except RuntimeError as err:
+        print(err)
+
+# Tried to refine an existing name to a different name
+catch_error(lambda: imgs.refine_names('batch', 'channel', 'height', 'width'))
+
+######################################################################
+# Most simple operations propagate names. The ultimate goal for named tensors is
+# for all operations to propagate names in a reasonable, intuitive manner. Many
+# common operations have been added at the time of the 1.3 release; here,
+# for example, is ``.abs()``:
+
+named_imgs = imgs.refine_names('N', 'C', 'H', 'W')
+print(named_imgs.abs().names)
+
+######################################################################
+# Accessors and Reduction
+# -----------------------
+#
+# One can use dimension names to refer to dimensions instead of the positional
+# dimension. These operations also propagate names. Indexing (basic and
+# advanced) has not been implemented yet but is on the roadmap. Using the ``named_imgs``
+# tensor from above, we can do:
+
+output = named_imgs.sum(['C'])  # Perform a sum over the channel dimension
+print(output.names)
+
+img0 = named_imgs.select('N', 0)  # get one image
+print(img0.names)
+
+######################################################################
+# Name inference
+# --------------
+#
+# Names are propagated on operations in a two step process called **name inference**.
+# The two steps are:
+#
+# 1. **Check names**: an operator may perform automatic checks at runtime that
+#    check that certain dimension names must match.
+# 2. **Propagate names**: name inference propagates output names to output tensors.
+#
+# Let's go through the very small example of adding 2 one-dim tensors with no
+# broadcasting.
+
+x = torch.randn(3, names=('X',))
+y = torch.randn(3)
+z = torch.randn(3, names=('Z',))
+
+######################################################################
+# **Check names**: first, we will check whether the names of these two tensors
+# match. Two names match if and only if they are equal (string equality) or at
+# least one is ``None`` (``None`` is essentially a special wildcard name).
+# The only one of these three that will error, therefore, is ``x + z``:
+
+catch_error(lambda: x + z)
+
+######################################################################
+# **Propagate names**: `unify` the two names by returning the most refined name of
+# the two. With ``x + y``,  ``X`` is more refined than ``None``.
+
+print((x + y).names)
+
+######################################################################
+# Most name inference rules are straightforward but some of them can have
+# unexpected semantics. Let's go through a couple you're likely to encounter:
+# broadcasting and matrix multiply.
+#
+# **Broadcasting**

```

Comment:
In markdown, one can use "^^^^^^^^^^^" as a smaller heading I think. It's really easy to confuse this heading with the bold "propagate names" and "check names" above.

Reply:
Thanks, I did not know this!!